test/javascripts/turbolinks_test.coffee

```diff
@@ -1,210 +1,946 @@
+          done()
+
+  describe 'with partial page replacement', ->
+    it 'script tags are evaluated when they are the subject of a partial replace', (done) ->
+      window.globalStub = stub()
+      visit url: 'inlineScriptInBody', options: {partialReplace: true, onlyKeys: ['turbo-area']}, ->
+        assert(globalStub.calledOnce, 'Script tag was not evaluated :(')
+        done()
+
+    it 'calls a user-supplied callback', (done) ->
+      yourCallback = stub()
+      visit url: 'noScriptsOrLinkInHead', options: {partialReplace: true, onlyKeys: ['turbo-area'], callback: yourCallback}, ->
+        assert(yourCallback.calledOnce, 'Callback was not called.')
+        done()
+
+    it 'script tags are not evaluated if they have [data-turbolinks-eval="false"]', (done) ->
+      window.globalStub = stub()
+      visit url: 'inlineScriptInBodyTurbolinksEvalFalse', options: {partialReplace: true, onlyKeys: ['turbo-area']}, ->
+        assert.equal(0, globalStub.callCount)
+        done()
+
+    it 'pushes state onto history stack', (done) ->
+      visit url: 'noScriptsOrLinkInHead',  options: {partialReplace: true, onlyKeys: ['turbo-area']}, ->
+        assert(Turbolinks.pushState.called, 'pushState not called')
+        assert(Turbolinks.pushState.calledWith({turbolinks: true, url: urlFor('/noScriptsOrLinkInHead')}, '', urlFor('/noScriptsOrLinkInHead')), 'Turbolinks.pushState not called with proper args')
+        assert.equal(0, replaceStateStub.callCount)
+        assert.equal(1, server.requests.length)
+        done()
+
+    it 'doesn\'t add a new history entry if updatePushState is false', (done) ->
+      visit url: 'noScriptsOrLinkInHead', options: {partialReplace: true, onlyKeys: ['turbo-area'], updatePushState: false}, ->
+        assert(pushStateStub.notCalled, 'pushState should not be called')
+        assert(replaceStateStub.notCalled, 'replaceState should be called')
+        done()
+
+    it 'uses just the part of the response body we supply', (done) ->
+      visit url: 'noScriptsOrLinkInHead',  options: {partialReplace: true, onlyKeys: ['turbo-area']}, ->
+        assert.equal("Hi there!", document.title)
+        assert.equal(-1, document.body.textContent.indexOf("YOLO"), 'Inserted more than just specified section.')
+        assert(document.body.textContent.indexOf("Hi bob") > -1, 'Hi bob is missing!')
+        done()
+
+    it 'triggers the page:load event with a list of nodes that are new (freshly replaced)', (done) ->
+      visit url: 'noScriptsOrLinkInHead', options: {partialReplace: true, onlyKeys: ['turbo-area']}, (event) ->
+        ev = event.originalEvent
+        assert.equal(true, ev.data instanceof Array)
+        assert.equal(1, ev.data.length)
+        node = ev.data[0]
+
+        assert.equal('turbo-area', node.id)
+        assert.equal('turbo-area', node.getAttribute('refresh'))
+        done()
+
+    it 'does not trigger the page:before-partial-replace event more than once', (done) ->
```

Comment:
I don't really know why this test is here except as clarification to how many times that event is fired by the library? I just ported it over :P